Add readme (#9)
canonical/edgex-checkbox-provider: README.md

# edgex-checkbox-provider

[![GitHub Latest Dev Tag)](https://img.shields.io/github/v/tag/canonical/edgex-checkbox-provider?include_prereleases&sort=semver&label=latest-dev)](https://github.com/canonical/edgex-checkbox-provider/tags) 
[![GitHub License](https://img.shields.io/github/license/canoical/edgex-checkbox-provider)](https://choosealicense.com/licenses/apache-2.0/) 
[![GitHub Pull Requests](https://img.shields.io/github/issues-pr-raw/canonical/edgex-checkbox-provider)](https://github.com/edgexfoundry/edgex-cli/pulls) 
[![GitHub Contributors](https://img.shields.io/github/contributors/canonical/edgex-checkbox-provider)](https://github.com/edgexfoundry/edgex-cli/contributors) 
[![GitHub Committers](https://img.shields.io/badge/team-committers-green)](https://github.com/orgs/canonical/teams/edgex/members) 

> [![GitHub Commit Activity](https://img.shields.io/github/commit-activity/m/edgexfoundry/edgex-cli)](https://github.com/canonical/edgex-checkbox-provider/commits)

Reviewer: I think many of the badges are redundant. All of tag, license, pull req, contributors, committers, commit activity are already available on github.

Author: Removed.

When edgexfoundry snap is released to the latest/beta channel, the corresponding checkbox test will be triggered. 

## Hardware

This table shows hardware used for edgexfoundry snap checkbox tests:
| Architecture | amd64              | amd64                          | amd64                  | arm64                         | arm64                       | arm64            | arm64                                     |           arm64           |
| ------------ | ------------------ | ------------------------------ | ---------------------- | ----------------------------- | --------------------------- | ---------------- | ----------------------------------------- | :-----------------------: |
| Model        | Tulip 17 BDW       | Caracalla                      | St. Louis              | RPI4B4G                       | RPI4B8G                     | RPI400           | CM4L                                      |        DragonBoard        |
| Information  | Dell Inspiron 5758 | Dell Edge Gateway 3003 [Media] | Dell Edge Gateway 5000 | armv7l Raspberry Pi 4 Model B | Raspberry  Pi 4 Model B 8GB | Raspberry Pi 400 | Raspberry Pi Compute Module 4 Lite - CM4L | Qualcomm DragonBoard 410c |

> | Ubuntu OS    | Core 16/16 LTS     | Core 16                        | Core 16                | 16 LTS                        | 16 LTS                      | 16 LTS           | 16 LTS                                    |          16 LTS           |

Reviewer: I'm not sure if this information should be in the source code. Is this hardware configured in the source?

Author: No. I think they are configured by checkbox team. Would you like me to remove this table?

Reviewer: Yes, I think it is better to remove.

Author: Removed. Thanks.

## Usage

### Run tests using checkbox CLI
```bash

> git clone https://github.com/canonical/edgex-checkbox-provider.git

Reviewer: This isn't necessary. It would be invalid in forks and github already has this.

Author: Removed.

cd edgex-checkbox-provider/
sudo ./manage.py install
checkbox-cli
```
![image-checkbox-test-plans](./assets/checkbox-test-plans.png)

Press SPACE to select an EdgeX test plan, then choose tests to run on your system:
![image-edgex-test-plan](./assets/edgex-test-plan.png)

### Run tests locally
Remove these snaps to avoid missing edgex-secretstore-token issue, if applicable:
```bash
snap remove --purge <edegx-device-service>
snap remove --purge <edgex-app-service>
```
Get the edgex-checkbox-provider:
```bash

> git clone https://github.com/canonical/edgex-checkbox-provider.git

Reviewer: The git clone command isn't necessary. It would be invalid in forks and github already has this.

Author: Removed.

```
```bash
# run all tests using locally built snaps
sudo ./run-all-tests-locally.sh-s edgexfoundry_3375.snap
```
```bash
# run a single test using snap from latest/beta channel
sudo ./run-all-tests-locally.sh -t test-rules-engine.sh
```
```bash
# run all tests using snap from latest/beta channel
sudo ./run-all-tests-locally.sh
```
```bash
# run tests showing output of tests even if passed
sudo ./run-all-tests-locally.sh -t -v
sudo ./run-all-tests-locally.sh -t test-install.sh -v
```
```bash
# get a list of available options
sudo ./run-all-tests-locally.sh -h
```

## Testing coverage
- Test installation of EdgeX snap
- Test security services proxy certs work properly
- Test that all services can be started properly
- Test that installing uses "current" based file paths in config files
- Test that refreshing uses "current" based file paths in config files
- Test that services start after refreshing 
- Test that services start after refreshing to this revision from self
- Test that the system management agent works with the snap
- Test that services are not listening on external network interfaces
- Test that the rules engine works with the snap

> - mandatory tests: interface, meminfo, connections ,[etc](https://github.com/canonical/edgex-checkbox-provider/blob/master/units/test-plan.pxu#L113).

Author: Change this to uppercase

```bash
snap remove --purge <edgex-device-service>
snap remove --purge <edgex-app-service>
```
Enter the desired test directory:
```bash
cd edgex-checkbox-provider/data/jakarta
```
Download a specific version snap:
```bash
snap download edgexfoundry <channel>
sudo ./run-all-tests-locally.sh -s <edgexfoundry.snap> -t <single test>
```
Here are some examples:

```bash
# run a single test using locally built snaps
snap download edgexfoundry --channel=2.1/stable
sudo ./run-all-tests-locally.sh -s edgexfoundry_3375.snap -t test-rules-engine.sh
```
```bash
# run all tests using locally built snaps
sudo ./run-all-tests-locally.sh-s edgexfoundry_3375.snap
```
```bash
# run a single test using snap from latest/beta channel
sudo ./run-all-tests-locally.sh -t test-rules-engine.sh
```
```bash
# run all tests using snap from latest/beta channel
sudo ./run-all-tests-locally.sh
```
```bash
# run tests showing output of tests even if passed
sudo ./run-all-tests-locally.sh -t -v
sudo ./run-all-tests-locally.sh -t test-install.sh -v
```
```bash
# get a list of available options
sudo ./run-all-tests-locally.sh -h
```

## Testing coverage

> - Test installation of EdgeX snap
> - Test security services proxy certs work properly
> - Test that all services can be started properly
> - Test that installing uses "current" based file paths in config files
> - Test that refreshing uses "current" based file paths in config files
> - Test that services start after refreshing 
> - Test that services start after refreshing to this revision from self
> - Test that the system management agent works with the snap
> - Test that services are not listening on external network interfaces
> - Test that the rules engine works with the snap

Reviewer: ```suggestion
- Test the installation of edgexfoundry snap
- Test security services proxy certs work properly
- Test that all services can be started properly
- Test that config paths don't include previous snap revision after installation
- Test that config paths don't include previous snap revision after refresh
- Test that services start after refreshing
- Test that services start after refreshing to the same revision
- Test that the system management agent works with the snap
- Test that services are not listening on external network interfaces
- Test that the rules engine works with the snap
```

# edgex-checkbox-provider

> ## Introduction

Reviewer: ```suggestion
```

> This project contains the checkbox tests of the [edgex foundry](https://docs.edgexfoundry.org/) snaps. 

Reviewer: ```suggestion
This project contains the checkbox tests of the [Edgex Foundry](https://docs.edgexfoundry.org/) snaps.
```

[Checkbox](https://checkbox.readthedocs.io/en/latest/) is a test automation software performed by the Canonical certification team. 

> The [upstream repository](/../../) is hosted on Github; 

Reviewer: `/../../` this should be an absolute URL to direct people looking at Launchpad mirror to Github.

> The [upstream repository](/../../) is hosted on Github; 

Reviewer: `/../../` this should be an absolute URL to direct people looking at Launchpad mirror to Github.

I suggest keeping the original inline URLs because Launchpad doesn't support markdown.

The tests run via the [Launchpad mirror](https://code.launchpad.net/checkbox-provider-edgex).

> When edgexfoundry snap is released to the latest/beta channel, the corresponding checkbox test will be triggered. 

Reviewer: ```suggestion
When edgexfoundry snap is released to a $TRACK/beta channel, the corresponding checkbox test will be triggered. 
```

## Usage
### Run tests using checkbox-edgexfoundry snap in isolation (recommend)
```bash
sudo snap install checkbox-edgexfoundry --devmode --channel=latest/edge

> ```

Reviewer: 1. Explanation should be in the body. E.g.: This is the recommended method as it runs all the tests in isolation.
2. Where to get it? -> https://snapcraft.io/checkbox-edgexfoundry
3. How is it built? -> https://launchpad.net/~ce-certification-qa/+snap/checkbox-edgexfoundry-edge
4. How should it be installed? -> Development Mode + example command

checkbox-edgexfoundry snap depends on extra environment variables, so setting test channel and release name:
```bash
sudo DEFAULT_TEST_CHANNEL=<"channel"> checkbox-edgexfoundry.<release name>
# Here is an example
sudo DEFAULT_TEST_CHANNEL="2.1/beta" checkbox-edgexfoundry.jakarta

```
### Run tests using checkbox CLI
```bash
cd edgex-checkbox-provider/
sudo ./manage.py install
checkbox-cli
```
![image-checkbox-test-plans](./assets/checkbox-test-plans.png)

Press SPACE to select an EdgeX test plan, then choose tests to run on your system:

> ![image-edgex-test-plan](./assets/edgex-test-plan.png)

Reviewer: Is it possible to place the information in textual form in a code block? It would be better for version control.

Reviewer: Scroll down and press SPACE to select the desired test plan:
```
 Select test plan
┌─────────────────────────────────────────────────┐
│    ( ) Dock Hot Plug tests                      │
│    ( ) EdgeX Fuji                               │
│    ( ) EdgeX Geneva                             │
│    ( ) EdgeX Hanoi                              │
│    (X) EdgeX Ireland                            │
│    ( ) EdgeX Jakarta                            │
│    ( ) Firewire tests                           │
└─────────────────────────────────────────────────┘
 Press <Enter> to continue                (H) Help
```

> ### Run tests locally

Reviewer: aren't all the tests instructions here to run them locally?

> ### Run tests locally

Reviewer: Please use short section to improve readability and ToC. Details can go under it.
- Usage
  - Run using checkbox-edgexfoundry snap
  - Run using Checkbox CLI
  - Run test scripts directly

Remove these snaps to avoid missing edgex-secretstore-token issue, if applicable:
```bash
snap remove --purge <edgex-device-service>
snap remove --purge <edgex-app-service>

> ```

Reviewer: How is this relevant? Some setup issue missing in tests?

> Get the edgex-checkbox-provider:

Reviewer: ```suggestion
Enter the desired test directory:
```

```bash
cd edgex-checkbox-provider/data/jakarta
```
Download a specific version snap:
```bash
snap download edgexfoundry <channel>
sudo ./run-all-tests-locally.sh -s <edgexfoundry.snap> -t <single test>
```
Here are some examples:

```bash
# run a single test using locally built snaps
snap download edgexfoundry --channel=2.1/stable
sudo ./run-all-tests-locally.sh -s edgexfoundry_3375.snap -t test-rules-engine.sh
```
```bash
# run all tests using locally built snaps
sudo ./run-all-tests-locally.sh-s edgexfoundry_3375.snap
```
```bash
# run a single test using snap from latest/beta channel
sudo ./run-all-tests-locally.sh -t test-rules-engine.sh
```
```bash
# run all tests using snap from latest/beta channel
sudo ./run-all-tests-locally.sh
```
```bash
# run tests showing output of tests even if passed
sudo ./run-all-tests-locally.sh -t -v
sudo ./run-all-tests-locally.sh -t test-install.sh -v
```
```bash
# get a list of available options
sudo ./run-all-tests-locally.sh -h

> ```

Reviewer: The last three lines are really enough, followed by one or two examples:

Get a list of available options:
```
sudo ./run-all-tests-locally.sh -h
```

For example, to run a single test with a local snap:
```bash
sudo ./run-all-tests-locally.sh -s edgexfoundry.snap -t test-rules-engine.sh
```

## Testing coverage
- Test installation of EdgeX snap
- Test security services proxy certs work properly
- Test that all services can be started properly
- Test that installing uses "current" based file paths in config files
- Test that refreshing uses "current" based file paths in config files
- Test that services start after refreshing 
- Test that services start after refreshing to this revision from self
- Test that the system management agent works with the snap
- Test that services are not listening on external network interfaces
- Test that the rules engine works with the snap

> - Mandatory tests: interface, meminfo, connections ,[etc](./units/test-plan.pxu#L113).

Reviewer: ```suggestion
- Mandatory tests: see [units/test-plan.pxu#L113](./units/test-plan.pxu#L113)
```

## Limitations

> - The current tests plan only covers [edgexfoundry snap](https://github.com/edgexfoundry/edgex-go/tree/main/snap), it does not cover [device services snaps](https://github.com/edgexfoundry/edgex-docs/blob/main/docs_src/getting-started/Ch-GettingStartedSnapUsers.md#device-service-snaps) and [app services snap](https://github.com/edgexfoundry/app-service-configurable/tree/main/snap)

Reviewer: ```suggestion
- The current tests plan only covers the [edgexfoundry](https://snapcraft.io/edgexfoundry) snap. It does not cover any of the device or app service snaps.
```

- edgex-secretstore-token content interface is not be covered by tests

## License

> [Apache-2.0](LICENSE)

Reviewer: Please remove this so that this PR doesn't have to wait for the LICENSE addition. It is not necessary to be here.
```suggestion
```
